add core dev 2023-09 (#276)
bitcointranscripts/bitcointranscripts: bitcoin-core-dev-tech/2023-09/p2p-working-session.md

---
title: P2P working session
categories: ['core-dev-tech']
tags: ['bitcoin core', 'p2p']
date: 2023-09-19
---

## Erlay

- Gleb is active and ready to move forward - [#21515](https://github.com/bitcoin/bitcoin/pull/21515)
- Are there people generally interested in review?

>   - I wanted first to convince myself that this is useful. I couldn't reproduce the numbers from the paper - 5% was what I got. 100 connections to erlay peers. My node is from a non-standard port. It may be that I don't have a normal sample. There is a pull request that could add RPC stats to bitcoind - that might get better numbers. We lose the stats once we disconnect.

Reviewer: ```suggestion
  - I wanted first to convince myself that this is useful. I couldn't [reproduce](https://github.com/vasild/bitcoin/wiki/Erlay) the numbers from the paper - 5% was what I got with ~100 connections. My node is listening on a non-standard port. It may be that I don't have a normal sample. There is a [pull request](https://github.com/bitcoin/bitcoin/pull/27534) that could add RPC stats to bitcoind - that might get better numbers. Current stats are per peer and we lose those stats once the peer disconnects.
```

>   - How much of INV traffic is redundant? Erlay claims to reduce the redundancy?

Reviewer: ```suggestion
  - How much of INV traffic is redundant? Erlay claims to reduce the redundancy.
```

  - It's the tool that I am providing. It looks at the certs.
- It makes it harder to make it consistent. It would be easier to run it against this data to compare against.
  - I can talk to the maintainer but it's not really their use case.
  - It may be worth trying.
  - also talked with Sjors about this. Docker then becomes a dependency. Another thing, when I put together this workshop - I'd like a clear distinction between the encoding and decoding code and data collection. I'd like to put the encoding/decoding tools into Bitcoin Core.
- How tightly the ASMAP code is tied to the bitcoin core project?
  - it is a necessity for us to use the ASMAP code
  - we already have some of the code in bitcoin core in python already.
  - sounds reasonable to me
- What problem is ASMAP solving?
  - replacing the bucketing logic of IPs (/16s) by autonomous systems
  - easier for attackers to get in the outbound rotation, slash 16 IPV4 with a few exceptions. ASMAP in general is being able to customize that. The data is uses autonomous systems and with bottleneck analysis where we can make improvements. Inbound evictions, addrman bucket, and 10 connections uses IPs

>   - I took from my addrman and used an ASMap and saw that 8 or 9 autonomous systems is what I saw. Saw that the random selection is always more than 6 autonomous systems.

Reviewer: ```suggestion
  - I took all addresses from my addrman, chose 10 random ones with different /16 prefix and checked in how many distinct ASs they belong - it was 8 or 9,always more than 6 autonomous systems.
```

  - is this worth spending time on. An attacker that only has access to Google cloud or AWS, how easy would it be to do?
  - you should run it when a different /16 with an empty addrman. You already have data and if you take out that restriction.
  - this will influence on the connections of the network (as it grows). ASs are one approximation for control. How good we don't know, but better than slash 16s. AWS probably has double digits AS numbers and thousands of IP ranges.
  - botnets have different looking IPs in many different ranges. Scattered.
  - If you could do a botnet and the network grew and then an attacker could position themselves to get more connections (Virtu presentation described effects of scaled up network with constant shares between ASs)

>   - In the case of 100x scale-up, random selection might be better than ASMap selection if [virtu's research](https://cryptic.to/advancing-bitcoin-2023/) is a concern. /16s make no difference.

Reviewer: ```suggestion
  - In the case of 100x scale-up, random selection might be better than ASMap selection if [virtu's research](https://cryptic.to/advancing-bitcoin-2023/) is a correct. /16s make no difference.
```

- AWS has multiple AS numbers, this softens the effect virtu described
- we could choose to merge all of AWS into a virtual AS. The question is where do we get the data from. I don't know where the place to have this sort of discussion apart from here. Does it belong on an issue or the mailing list?
- before it is included in a release, it makes sense to post it on the mailing list.

## Broadcasting out all txs through Tor (Vasil)

> - The idea is to broadcast local transactions from the wallet or RPC - right now we broadcast to all peers, which makes it easy to correlate txs to IP addresses.

Reviewer: ```suggestion
- The idea is to change how we broadcast local transactions from the wallet or RPC - right now we broadcast to all peers, which makes it easy to correlate txs to IP addresses.
```

> - Only broadcast to Tor peers if possible. Send to someone random and disconnect. If I only tell Tor peers and a safety net to tell if it is propagated and if not broadcast.

Reviewer: ```suggestion
- Only broadcast to Tor peers if possible. Send to someone random and disconnect. If I only tell Tor peers add a safety net to tell if it is propagated and if not broadcast to all.
```

- Can see the timing and where txs originate from.
- The trickling mechanism adds randomness and doesn't obscure things from an attacker that sees everything
- Inspired by [Linking Lion](https://b10c.me/observations/05-inbound-connection-flooder-down/)

> - Problem is the change and where to put that.

Reviewer: ```suggestion
```

- If we send it to a few peers over Tor and some of them are spies, they have a privileged position to see. If they come to our mempool and ask and they can confirm if it is yours. Then if some asks and then we pretend we don't have it.
  - you are changing the behavior of what we allow in the mempool.
  - If some sends an INV or sends GETDATA - parent and child issues also complicate things. Convinced I can hide it.

> - Now I'm leaning the other way, which is not to put it in the mempool. But where? Need to put it on disk. In the mempool there are unbroadcasted txs. The code is already there. If it doesn't come back, you need to put it somewhere.

Reviewer: ```suggestion
- Now I'm leaning the other way, which is to broadcast without putting it in the mempool. Need to store it on disk too before it as been received by the network, so that it does not get lost if the the node restarts. In the mempool there are unbroadcasted txs. The code is already there. If it doesn't come back, you need to store it safely for retry later.
```

- There are two cases to think about. It comes from your wallet or it's from RPC. What happens if the RPC comes from the wallet?

> - SendoutTx would put in your mempool. Subject to trickling mechanism will go in the ToBeINVed.

Reviewer: ```suggestion
- sendrawtransaction would put in your mempool. Subject to trickling mechanism will go in the ToBeINVed.
```

>   - What about announcing to some of your peers, wait to see it echoed back and then send to all.

Reviewer: ```suggestion
  - What about announcing to some of your peers, wait to see it echoed back and then send to all if it does not come back.
```

  - How bitcoin core worked until 0.5
  - It's not trivial I think. Need to act differently for that Txs than Txs coming from elsewhere which may also be a tell. And also need a more aggressive broadcast mechanism until it is echoed back.
  - we don't want to send it to everyone. Then we send through Tor, then you could skip all the rest and then you destroy the assumption of the attacker.
  - if you fluff it, then someone can assume that you are the source wallet. Even if none of the behavior changes, it destroys the assumption reliability. It lowers the confidence level.
  - chainalysis can't see where it originates
- if you have a time-sensitive protocol, that might be an issue
- after 20 min try the same
  - 20 min seems enormous. Should only be a few mins. An attacker watching all the connections. If you only send it to 1 then they can't do the timing connection. It would be nice as a stretch goal, to obscure that the mechanism is being used.

>   - In addition to Tor peers, send to clearnet connections through Tor. In I2P, no source connections.

Reviewer: ```suggestion
  - In addition to Tor peers, send to clearnet connections through Tor. In I2P, use transient source addresses.
```

- can you coin flip this, by coinflip
- if you do it twice then they can tell it is yours
- what do you think about sending out once immediately and after a short delay (seconds) do it again.

> - This happens now because we have only 2 threads for the connection manager. So it is going to take some time. Should I broadcast it to 5 unconditionally or 1 and as soon as

Reviewer: ```suggestion
- This happens now because we have only 2 threads for the connection manager. So it is going to take some time. Should I broadcast it to 5 unconditionally or 1, then another one and then another one and stop as soon as the transaction is echoed back.
```

- do we broadcast after we see it?

> - we broadcast like we've never seen it before. Will remove from the mempool and

Reviewer: ```suggestion
```

- a goal is to treat our transactions like others. By keeping it out of the mempool, that goal is achieved.

> - so we will keep out the mempool

Reviewer: ```suggestion
- so we will keep it out of the mempool
```

- and then we can think about wallet interaction
- not going to see 0 balance, but now we will see the same balance

> - can't do this by default until we have the wallet interaction in place. And we deal with the case that someone makes a dependent tx because then we use the privacy advantages. The correction solution is a new state in the wallet for transactions that can't be spent.

Reviewer: ```suggestion
- can't do this by default until we have the wallet interaction in place. And we deal with the case that someone makes a dependent tx because then we use the privacy advantages. The correct solution is a new state in the wallet for transactions that can't be spent.
```

> - start with an unstable RPC

Reviewer: ```suggestion
- start with an the RPC
```

> - when you send it over RPC it will return immediately to you and then start sending to the network. When it returns, it means seen by the network.

Reviewer: ```suggestion
- when you send it over RPC, the RPC will return immediately and then start sending the transaction to the network. When the RPC returns, it does not mean that the transaction is seen by the network.
```

- could do wait to broadcast or something like that.

> - new RPC, would not overload `sendrawTx`. Velocity risk is when you have to explain questions about it. Can call it an experimental API and then roll it in or include in `sendrawTX`. Don't return the PRC as a string but as a JSON object. Everything should be an object.

Reviewer: ```suggestion
- new RPC, would not overload `sendrawtransaction`. Can call it an experimental API and then roll it in or include in `sendrawtransaction`. Don't return the PRC as a string but as a JSON object. Everything should be an object.
```

- Introduced as two block-only connections, always with the idea we wanted a bigger number
  - Having more connections that isn't a resource drain
  - Have had other work, like `addrRelay`, network stuff to prep for this
    - You cannot know whether you will send inbound connections but it also doesn't matter because validation is cheap. It's the sending of Txs are expensive.
    - What is important is predicting whether or not the peer wants to you to send TXs
  - Increasing the number of default outbound block-only conns, we propose 8 block-only connections
  - Needs to increase the number of inbound slots available. Not just increase the number, but add additional slots that would only be used for inbound block-only traffic. (130-140 slots)
  - Main open question is about the ratio
- If you don't offer the bloomfilter, you know it's blockrelay or another implementation
  - We send this flag to our peers, but if we are in blocks only mode and if we send rawtx and we can still txs to our peers unless they have already sent
- There is no way to say, I'm not sending your transactions
  - The relay flag in the version message.

>   - If I send the set the `fRely` flag false, I cannot deduce will they ever send me a tx doesn't seem possible

Reviewer: ```suggestion
  - If I send the set the `fRelay` flag false, I cannot deduce will they ever send me a tx doesn't seem possible
```

- Graphml that has it running on docker compose.
- Zipkin took the test framework class, self.nodes is a docket containers. Can copy open existing tests. These are called scenarios.
- Scenario in a loop: a mining scenario. Or could also execute once and shut down.
- Transaction scenarios that bring the network to life and then you could run your scenarios on top of that.
- Network set up is deterministic and the scenario is deterministic.
- Log aggregation needs the most work and needs the most input:
- Prometheus: we are running a graphana service, prometheus service and fork monitor.
- Ideally we have all the logs aggregated somewhere. Needs
- A p2p monitor would be something I'd like to put in there.
- Have a dashboard where you can see what is happening.
- WarnetCLI can going into a specific container or run RPC command
- be able to track how a message went from one place to another
Could be a tx tracker that travels from node to node
- How many can you run on a laptop?
- we could do 0-50 nodes on your laptop or you want to do something that is long running figure out where else to put it

> I've put up 500 on a dev box

Reviewer 2: ```suggestion
- I've put up 500 on a dev box
```

> Could also attract academics and give them some tooling to test a PR. Even testing an attack. Could deploy that.

Reviewer 2: ```suggestion
- Could also attract academics and give them some tooling to test a PR. Even testing an attack. Could deploy that.
```
